src/repositories/gitlab_updater.rs

```diff
@@ -0,0 +1,323 @@
+use crate::error::Result;
+use crate::{db::Pool, Config};
+use chrono::{DateTime, Utc};
+use log::{info, warn};
+use postgres::Client;
+use reqwest::{
+    blocking::Client as HttpClient,
+    header::{HeaderMap, HeaderValue, ACCEPT, AUTHORIZATION, USER_AGENT},
+};
+use serde::Deserialize;
+use std::sync::Arc;
+
+use crate::repositories::{Updater, APP_USER_AGENT};
+
+const GRAPHQL_UPDATE: &str = "query($ids: [ID!]!) {
+    projects(ids: $ids) {
+        nodes {
+            id
+            nameWithNamespace
+            lastActivityAt
+            description
+            starCount
+            forksCount
+            openIssuesCount
+        }
+    }
+}";
+
+const GRAPHQL_SINGLE: &str = "query($fullPath: ID!) {
+    project(fullPath: $fullPath) {
+        id
+        nameWithNamespace
+        lastActivityAt
+        description
+        starCount
+        forksCount
+        openIssuesCount
+    }
+}";
+
+/// How many repositories to update in a single chunk.
+const UPDATE_CHUNK_SIZE: usize = 5;
+
+pub struct GitlabUpdater {
+    client: HttpClient,
+    pool: Pool,
+}
+
+impl Updater for GitlabUpdater {
+    /// Returns `Err` if the access token has invalid syntax (but *not* if it isn't authorized).
+    /// Returns `Ok(None)` if there is no access token.
+    fn new(config: Arc<Config>, pool: Pool) -> Result<Option<Self>> {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+
+        if let Some(token) = &config.gitlab_accesstoken {
+            headers.insert(
+                AUTHORIZATION,
+                HeaderValue::from_str(&format!("Bearer {}", token))?,
+            );
+        } else {
+            warn!("will try to retrieve Gitlab stats without token since none was provided");
+        }
+
+        let client = HttpClient::builder().default_headers(headers).build()?;
+
+        Ok(Some(GitlabUpdater { client, pool }))
+    }
+
+    fn load_repository(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match Self::repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        let project_path = format!("{}/{}", name.owner, name.repo);
+        // Avoid querying the Gitlab API for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&project_path, &name.host],
+        )? {
+            return Ok(Some(row.get("id")));
+        }
+
+        // Fetch the latest information from the Gitlab API.
+        let response: GraphResponse<GraphProjectNode> = self.graphql(
```

Comment:
Also coming from a backfill, I saw some error (likely) originating here somewhere: 
```
error decoding response body: invalid type: null, expected i64 at line 1 column xxx
```

for (at least) the following repo URLs: 
https://gitlab.com/musoke/libinspire
https://gitlab.com/barry.van.acker/phreak-rs
https://gitlab.com/chipaltman/markdust
https://gitlab.com/valeth/javelin.git
https://gitlab.com/ahmedcharles/mct
https://gitlab.com/juggle-tux/farbfeld-rs
https://gitlab.com/ahmedcharles/beads
https://gitlab.com/dgibson/abstractgraph.git
https://gitlab.com/storedbox/cargo-yaml.git

(I canceled the backfill then)

The error could be reproduced on the same crate when retrying, also deleting these releases from my `releases` table did allow the backfill to continue. So it doesn't seem to be rate-limiting.

Reply:
(the repos I saw do exist)

Reply:
The issue was that "open issues count" can be null. Fixed.